### DISCUSSION TITLE
Approach to writing End-to-End tests for flows that require Action Scheduler

### DISCUSSION THREAD

**Discussion by Sidsector9** [woocommerce/woocommerce#40664] [General]
Hello, we am working on a Woo Payment Gateway plugin that does product sync through Action Scheduler.
We have set up Playwright to write end-to-end tests and were wondering what is the best approach to write E2E tests for flows that require Action Scheduler without facing timeouts?

I dug into the Woo Core tests and couldn't find much. Does Woo have a general approach to handle these flows?

**Comment by jeffpaul** [comment]
@woocommerce/solaris curious if anyone from your team has insights on this perhaps?

**Comment by lanej0** [comment]
Hi @Sidsector9 and @jeffpaul -- I'm actually trying to chase this down myself right now.  I've written some new end-to-end tests for the Analytics section, which requires some Action Scheduler events to run to populate the data.

Currently, I'm [forcing them through the queue](https://github.com/woocommerce/woocommerce/pull/40504/files) in the `beforeEach()` hook -- but I'm not super happy with that approach, because it's kind of slow.  I'm going to try asking around to a few folks and see if there's a better/easier/faster approach that I've overlooked.  I'll report back once I find out!

**Reply by Sidsector9** [reply]
Thanks @lanej0! I had something similar in mind but time is the concern. I'll give it a try and see how long it takes.

**Comment by lanej0** [comment]
Hi folks, one of the devs in the core team helped me out with putting together a plugin that can be called to immediately process the Action Scheduler queue.  The PR where that's being used is here:

https://github.com/woocommerce/woocommerce/pull/40504

I'm not sure if you're using `wp-env` like in the core test suite for running your tests, but if you are, you can just install the plugin during setup.  Then, there's a simple API call to process the AS queue.  I found that I had to introduce a 5 second delay before the processing was complete -- but doing that one time was much faster than the 30 seconds+ it was taking when I was manually looping through the queue in the UI.

Hope that helps!

**Reply by Sidsector9** [reply]
We're using `wp-env` and this plugin will be helpful to reduce the wait time. Thanks @lanej0 this was a big help!
